### PR TITLE
Allow the api name to appear in tags instead of the metric name

### PR DESCRIPTION
**Pull request by kbacha**
### Summary

Data dog prefers to use tags to query and divide statistics up by.
By putting the api name in the key, we cannot create dashboards that
are keyed off the api name. This is accomplished via a new plugin
setting called `tag_api_name`. It is a boolean field that when set
to true will remove the api_name from the metric key and add it as
a tag. It will append this tag to whatever tags are defined by the
metric configuration.

### Full changelog

* Implemented schema changes
* Implemented test for api name inclusion in key set
* Implemented appended api name to key set
* Refactored plugin to help readability.

### CODE REVIEW

**Comment by thibaultcha** [comment]
Thanks for the contribution! We will take a look at it soon.

File: kong/plugins/datadog/handler.lua [hunk 1]
```diff
@@ -2,143 +2,158 @@ local BasePlugin       = require "kong.plugins.base_plugin"
...
-  custom_id   = function(consumer)
-    return consumer and consumer.custom_id
-  end,
-  username    = function(consumer)
-    return consumer and consumer.username
-  end
+local ConsumerId = {
+  consumer_id = function(consumer) return consumer and string_gsub(consumer.id, "-", "_") end,
+  custom_id   = function(consumer) return consumer and consumer.custom_id end,
+  username    = function(consumer) return consumer and consumer.username end
```
Comment (shashiranjan84) [inline kong/plugins/datadog/handler.lua]: I don't see any reason to make this change and we do not use camelcase for naming variables.

Comment (kbacha) [reply kong/plugins/datadog/handler.lua]: I was following what appeared to be lua conventions above for `DatadogHandler` and `BasePlugin`. This table works as a module in this file so I wanted to use camelcase to demonstrate that.

**Comment by shashiranjan84** [comment]
@kbacha thanks for contribution. Plugin was designed as because of some business reasons. I have some questions regarding PR,

> By putting the api name in the key, we cannot create dashboards that
are keyed off the api name

Are you trying to aggregate the metrics for all API and use tag as filter?

**Comment by kbacha** [comment]
@shashiranjan84 Yes, that's exactly what we are using it for. The tags allow us to use template variables on dashboards to filter and histogram different metrics.

We made this behavior optional so it would not change the current default behavior.

**Comment by kbacha** [comment]
@shashiranjan84 Are there more changes you'd like me to make?

**Comment by shashiranjan84** [comment]
@kbacha I did not get time to review it properly due to holidays and other work. On high level logic seems to be fine but I do see some changes you made was not required. I will leave further comment once I have some time to look into it further, stay tuned.

**Comment by kbacha** [comment]
@shashiranjan84 Ping?

File: kong/plugins/datadog/handler.lua [hunk 2]
```diff
@@ -2,143 +2,158 @@ local BasePlugin       = require "kong.plugins.base_plugin"
...
 }
 
+local function get_consumer_id(metric_config, message)
+  local consumer_id_func = consumer_id_functions[metric_config.consumer_identifier]
+  return consumer_id_func(message.consumer)
+end
+
+local metric_functions = {}
```
Comment (shashiranjan84) [inline kong/plugins/datadog/handler.lua]: Again unnecessary change, we know all the functions of this table and constructor should include all those functions. Like it was done before. I also don't see any reason to change the name.

File: kong/plugins/datadog/handler.lua [hunk 3]
```diff
@@ -2,143 +2,158 @@ local BasePlugin       = require "kong.plugins.base_plugin"
...
+  end
 
   local logger, err = statsd_logger:new(conf)
   if err then
     ngx_log(NGX_ERR, "failed to create Statsd logger: ", err)
     return
   end
 
+  local api_name    = string_gsub(message.api.name, "%.", "_")
+  local name_prefix = build_name_prefix(api_name, conf)
```
Comment (shashiranjan84) [inline kong/plugins/datadog/handler.lua]: I would prefer 
```
local name_prefix = conf.tag_api_name and "" or "."
```

over doing a context switch.

File: kong/plugins/datadog/handler.lua [hunk 4]
```diff
@@ -2,143 +2,158 @@ local BasePlugin       = require "kong.plugins.base_plugin"
...
+  local stat_name   = build_stat_names(name_prefix)
```
Comment (shashiranjan84) [inline kong/plugins/datadog/handler.lua]: Again we know the content of stat_name table, no need to add new function.

File: kong/plugins/datadog/handler.lua [hunk 5]
```diff
@@ -2,143 +2,158 @@ local BasePlugin       = require "kong.plugins.base_plugin"
...
+  local stat_value  = collect_stat_values(message)
```
Comment (shashiranjan84) [inline kong/plugins/datadog/handler.lua]: ditto

File: kong/plugins/datadog/handler.lua [hunk 6]
```diff
@@ -2,143 +2,158 @@ local BasePlugin       = require "kong.plugins.base_plugin"
...
+
   for _, metric_config in pairs(conf.metrics) do
-    local metric = metrics[metric_config.name]
+    local metric_func = metric_functions[metric_config.name]
 
-    if metric then
-      metric(api_name, message, metric_config, logger)
+    local tags = metric_config.tags
+    if conf.tag_api_name then
+      tags = merge_tags(tags, api_name)
```
Comment (shashiranjan84) [inline kong/plugins/datadog/handler.lua]:
```
local tags
 if conf.tag_api_name then
    tags = metric_config.tags or {}
    table_insert(tags, api_name)
 end
```

**Comment by kbacha** [comment]
@shashiranjan84 Thanks, I'll get to work adjusting the style.

**Comment by kbacha** [comment]
@shashiranjan84 Take another look. I've tried to clean all the unnecessary changes up. Sorry, I'm new to lua and was coding with different conventions in mind.

File: kong/plugins/datadog/handler.lua [hunk 7]
```diff
@@ -25,59 +25,52 @@ local get_consumer_id = {
...
+  end,
 }
 
 
 local metrics = {
-  status_count = function (api_name, message, metric_config, logger)
-    local fmt = string_format("%s.request.status", api_name,
-                       message.response.status)
+  status_count = function(name_prefix, message, metric_config, logger, tags)
+    local fmt = string_format("%srequest.status", name_prefix, message.response.status)
```
Comment (hbagdi) [inline kong/plugins/datadog/handler.lua]: Please follow the code style.
https://github.com/Kong/kong/blob/master/CONTRIBUTING.md#code-style

Please take care of this all over the patch.

File: kong/plugins/datadog/handler.lua [hunk 8]
```diff
@@ -25,59 +25,52 @@ local get_consumer_id = {
...
 
     logger:send_statsd(string_format("%s.%s", fmt, message.response.status),
-                       1, logger.stat_types.counter,
-                       metric_config.sample_rate, metric_config.tags)
+                       1, logger.stat_types.counter, metric_config.sample_rate, tags)
```
Comment (hbagdi) [inline kong/plugins/datadog/handler.lua]: Please follow the code style.
https://github.com/Kong/kong/blob/master/CONTRIBUTING.md#code-style

Please take care of this all over the patch.

Also, avoid unnecessary changes to keep the patch readable.

File: kong/plugins/datadog/handler.lua [hunk 9]
```diff
@@ -25,59 +25,52 @@ local get_consumer_id = {
   end,
   username    = function(consumer)
     return consumer and consumer.username
-  end
+  end,
```
Comment (hbagdi) [inline kong/plugins/datadog/handler.lua]: Avoid unnecessary changes in the patch..

File: kong/plugins/datadog/handler.lua [hunk 10]
```diff
@@ -88,16 +81,23 @@ local function log(premature, conf, message)
...
+  local stat_name   = {
+    request_size     = name_prefix .. "request.size",
+    response_size    = name_prefix .. "response.size",
+    latency          = name_prefix .. "latency",
+    upstream_latency = name_prefix .. "upstream_latency",
+    kong_latency     = name_prefix .. "kong_latency",
+    request_count    = name_prefix .. "request.count",
   }
-  local stat_value = {
+  local stat_value  = {
```
Comment (hbagdi) [inline kong/plugins/datadog/handler.lua]: extra space

File: kong/plugins/datadog/handler.lua [hunk 11]
```diff
@@ -88,16 +81,23 @@ local function log(premature, conf, message)
...
 
-  local api_name   = string_gsub(message.api.name, "%.", "_")
-  local stat_name  = {
-    request_size     = api_name .. ".request.size",
-    response_size    = api_name .. ".response.size",
-    latency          = api_name .. ".latency",
-    upstream_latency = api_name .. ".upstream_latency",
-    kong_latency     = api_name .. ".kong_latency",
-    request_count    = api_name .. ".request.count",
+  local logger, err = statsd_logger:new(conf)
```
Comment (hbagdi) [inline kong/plugins/datadog/handler.lua]: Any reason for this change? 
Please avoid unnecessary changes.

File: kong/plugins/datadog/handler.lua [hunk 12]
```diff
@@ -106,25 +106,23 @@ local function log(premature, conf, message)
...
 
-  local logger, err = statsd_logger:new(conf)
-  if err then
-    ngx_log(NGX_ERR, "failed to create Statsd logger: ", err)
-    return
-  end
-
   for _, metric_config in pairs(conf.metrics) do
-    local metric = metrics[metric_config.name]
+    local metric_func = metrics[metric_config.name]
```
Comment (hbagdi) [inline kong/plugins/datadog/handler.lua]: since `metrics` is a table of functions it's implied that `metric` is a function. Please revert the name to `metric`.

File: spec/03-plugins/08-datadog/01-log_spec.lua [hunk 13]
```diff
@@ -287,6 +314,41 @@ describe("Plugin: datadog (log)", function()
     assert.contains("kong.dd3.latency:%d+|g|#T2:V2:V3,T4", gauges, true)
   end)
 
+  it("logs metrics with tags with api name in tag", function()
```
Comment (hbagdi) [inline spec/03-plugins/08-datadog/01-log_spec.lua]: can we add a test to ensure that api_name tag is not added when `tag_api_name`  is false?

File: kong/plugins/datadog/handler.lua [hunk 14]
```diff
@@ -25,59 +25,52 @@ local get_consumer_id = {
...
-      local stat = string_format("%s.user.uniques", api_name)
-
-      logger:send_statsd(stat, consumer_id, logger.stat_types.set,
-                         nil, metric_config.tags)
+      local stat = string_format("%suser.uniques", name_prefix)
+      logger:send_statsd(stat, consumer_id, logger.stat_types.set, nil, tags)
     end
   end,
-  request_per_user = function (api_name, message, metric_config, logger)
+
```
Comment (hbagdi) [inline kong/plugins/datadog/handler.lua]: avoid this line jump

**Comment by hbagdi** [comment]
@kbacha @shashiranjan84 
Do we even need a configurable `tag_api_name`? Would it hurt in any way if we add this tag and not provide a configurable option to the end user?

Also, should we consider adding a tag for `consumer_id` as well while we're at it?

File: kong/plugins/datadog/handler.lua (hunk 9 above)
Comment (kbacha) [reply kong/plugins/datadog/handler.lua]: @hbagdi This was in the style guide:
https://github.com/Kong/kong/blob/master/CONTRIBUTING.md#tables

I added the trailing space.

File: kong/plugins/datadog/handler.lua (hunk 11 above)
Comment (kbacha) [reply kong/plugins/datadog/handler.lua]: @hbagdi This avoid unnecessary memory allocations below if we can return early.

**Comment by kbacha** [comment]
@hbagdi I made it a configuration so that we didn't change the default behavior. Consumer id also seems reasonable but I'm not sure if there is a limit to the number of values that can be assigned. How many consumers do larger installations generally have?

File: spec/03-plugins/08-datadog/01-log_spec.lua (hunk 13 above)
Comment (kbacha) [reply spec/03-plugins/08-datadog/01-log_spec.lua]: Sure!

**Comment by kbacha** [comment]
@hbagdi @shashiranjan84 Ok, how is it now?

**Comment by hbagdi** [comment]
@kbacha Regarding number of consumers, I can easily imagine tens of thousands of consumers or more on a large installation. Even hundred thousand plus.

**Comment by kbacha** [comment]
@hbagdi Cool, if we implement a consumer id tag, might be worth making it a toggle just as I have it for api name now. In case it causes problems with datadog or causes someones costs to go up.

From their documentation: https://help.datadoghq.com/hc/en-us/articles/204312749-Getting-started-with-tags
> We store one time series per host + metric + tag combination on our backend, thus we cannot support infinitely bounded tags. Please don't include endlessly growing tags in your metrics, like timestamps or user ids. Please limit each metric to 1000 tags.

**Comment by kbacha** [comment]
@shashiranjan84 @hbagdi Can I get 👀 on again?

**Comment by kbacha** [comment]
@shashiranjan84 @hbagdi Still waiting on feedback

**Comment by shashiranjan84** [comment]
@kbacha sorry slammed with other work, I will try review it next week.

**Comment by shashiranjan84** [comment]
@kbacha LGTM, mind sending a PR for Doc too if you not already to https://github.com/Kong/getkong.org

**Comment by kbacha** [comment]
@shashiranjan84 Proposed: https://github.com/Kong/getkong.org/pull/607

**Comment by kbacha** [comment]
@shashiranjan84 I spent the day resolving merge conflicts. It's still failed on travis but the error looks unrelated to my changes.

File: kong/plugins/datadog/handler.lua [hunk 15]
```diff
@@ -128,25 +137,23 @@ local function log(premature, conf, message)
...
-    ngx_log(NGX_ERR, "failed to create Statsd logger: ", err)
-    return
-  end
-
   for _, metric_config in pairs(conf.metrics) do
     local metric = metrics[metric_config.name]
 
-    if metric then
-      metric(name, message, metric_config, logger)
+    local tags = metric_config.tags or {}
```
Comment (kbacha) [inline kong/plugins/datadog/handler.lua]: I did this at your request but I believe that the config can live from request to request and seems to build up. Is that true?

Comment (shashiranjan84) [reply kong/plugins/datadog/handler.lua]: Sorry, can you elaborate more?

Comment (kbacha) [reply kong/plugins/datadog/handler.lua]: Well, if there are tags on the config then this is just a reference to the tags on the config. It then gets modified when we insert the api name. If the reference isn't cleared between requests then every request will append more to the table.

Something i noticed when working locally but wasn't sure if i was interpreting it correctly.

Comment (shashiranjan84) [reply kong/plugins/datadog/handler.lua]: I see it now, but @thibaultcha correct me if I am wrong, a new copy of plugin config is pulled for each request from the cache, so making change to config in one request should not affect other request.

Comment (thibaultcha) [reply kong/plugins/datadog/handler.lua]: A plugin's `config` outlives a single requests - yes, it is shared between multiple ones, as it is cached in the Lua-land (and should be read-only, in the future).

Comment (shashiranjan84) [reply kong/plugins/datadog/handler.lua]: @kbacha mind updating the code to use copy of tag table? `kong.tools.utils` has a `shallow_copy` function.

Comment (kbacha) [reply kong/plugins/datadog/handler.lua]: For sure!

Comment (kbacha) [reply kong/plugins/datadog/handler.lua]: Added a test to demonstrate what happens and used shallow copy to correct it.

**Comment by kbacha** [comment]
@hbagdi Anything left to change your ❌ to a ✅

**Comment by AndrewFarley** [comment]
LOL, I re-implemented this because this has been pending for so long and I was disgusted at the lack-of-tags also.  At: #3465   Please merge this soon, datadog is unusable without this.

**Comment by AndrewFarley** [comment]
Actually @kbacha I was reviewing this merge... can I provide some further recommendations...?  Please see: https://github.com/Kong/kong/pull/3465/files in my merge, your merge here still has metrics `kong.request.status.200` which it shouldn't... that should be `kong.request.status` with a tag of `status_code:200`.  Could you please add that to this merge please?  Additionally, once that is done, as my merge implements, having a "count" or a "total" is useless, as in datadog you can aggregate the sum of the metric.

**Comment by coopr** [comment]
What is remaining to do to get this PR merged?

**Comment by AndrewFarley** [comment]
@coopr Honestly, my suggestion above can be done in a future MR.  Even what the author of this patch has done is already a step ahead, although imho this migration shouldn't be a "half-step" this should be a full measure of properly using tags.  I did this feature in my other MR which I closed, as mentioned above... but after weeks of attempting to use Kong and Datadog we actually ended up not using Datadog, so I don't have the billable time to try to help improve this PR, and nor do I have free time.  So, feel free to merge this.

**Comment by kikito** [comment]
Hi @kbacha , while we are very grateful for your PR, we finally decided to merge #5154 instead. It does changes similar to yours - the service id is *always* removed from the metric, instead of being "toggleable". It is also included as a tag, amongst other changes. Give it a look!
